### DISCUSSION TITLE
Entity discovery not working with bun after upgrade to 6.2.2

### DISCUSSION THREAD

**Discussion by co-sic** [mikro-orm/mikro-orm#5496] [General]
Hey, i think that #5464 broke folder-based entity discovery with bun. I know bun is not officially supported, so i'm not opening an issue. I don't have a reproduction yet and will just downgrade for now, but if you want to look into it, i can provide one tomorrow. 

Error:
```
[info] MikroORM version: 6.2.2
[discovery] ORM entity discovery started, using TsMorphMetadataProvider
[discovery] - processing 17 files
[discovery] - ignoring file index.ts
[discovery] - ignoring file index.ts
[discovery] - ignoring file ScopedPermissionType.ts
[discovery] - ignoring file Language.ts
[discovery] - ignoring file PermissionType.ts
[discovery] - ignoring file PermissionScope.ts
[discovery] - processing entity Role (~/src/entities/Role.ts)
119 |             this.initSourceFiles();
120 |         }
121 |         const source = this.sources.find(s => s.getFilePath().endsWith(core_1.Utils.stripRelativePath(tsPath)));
122 |         if (!source && validate) {
123 |             throw new core_1.MetadataError(`Source file '${tsPath}' not found. Check your 'entitiesTs' option and verify you have 'compilerOptions.declaration' enabled in your 'tsconfig.json'. If you are using webpack, see https://bit.ly/35pPDNn`);
124 |         }
^
MetadataError: Source file './src/entities/Role.ts' not found. Check your 'entitiesTs' option and verify you have 'compilerOptions.declaration' enabled in your 'tsconfig.json'. If you are using webpack, see https://bit.ly/35pPDNn
at getSourceFile (~/node_modules/@mikro-orm/reflection/TsMorphMetadataProvider.js:124:15)
at readTypeFromSource (~/node_modules/@mikro-orm/reflection/TsMorphMetadataProvider.js:96:110)
at initPropertyType (~/node_modules/@mikro-orm/reflection/TsMorphMetadataProvider.js:71:23)
at initProperties (~/node_modules/@mikro-orm/reflection/TsMorphMetadataProvider.js:32:38)
at loadEntityMetadata (~/node_modules/@mikro-orm/reflection/TsMorphMetadataProvider.js:16:33)
at discoverEntity (~/node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:373:99)
at ~/node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:278:9
at asyncFunctionResume (native:1:1)
at promiseReactionJobWithoutPromiseUnwrapAsyncContext (native:1:1)
at promiseReactionJob (native:1:1)
at processTicksAndRejections (native:1:1)
```

**Comment by B4nan** [comment]
I surely don't want people to downgrade, things that worked should remain working, so repro welcome :] Maybe it even won't be bun specific, let's see.

cc @xmajzel

**Comment by xmajzel** [comment]
I'm not sure why it shouldn't work, looks like it might be a specific case for `TsMorphMetadataProvider`, despite the tests for that provider working.

Certainly reproduction would be fine, what I was wondering  **what is the name of the file** compared to the **name of the class within the file** (@co-sic)?

When fixing https://github.com/mikro-orm/mikro-orm/pull/5464, omitting setting Global metadata caused the same error as provided by @co-sic in `TsMorphMetadataProvider.test.ts` test file _(See comment in code bellow)_.
- So I think the problem could theoretically be **what key** the metadata is set under for this entity in the global metadata.

**MetadataDiscovery.getSchema:**

```ts
 if (entity instanceof EntitySchema) {
      if (filepath) {
        // Try comment out following line. Then run TsMorphMetadataProvider.test.ts tests to see error.
        MetadataStorage.getMetadata(entity.meta.className, filepath);
      }

      return entity;
    }
```

cc @Ruby184

**Reply by co-sic** [reply]
I will provide a reproduction tomorrow :)

**Comment by co-sic** [comment]
@xmajzel here is the reproduction: https://github.com/co-sic/bun-playground/tree/mikro-orm . You should get the error there if you follow the README to start. If you downgrade to 6.2.1 and run again, it will work. Let me know if you need anything else

**Reply by Ruby184** [reply]
I have tried the reproduction, and I can confirm that it does not work using Bun but is working without any issue using Node. The problem seems to be in the global metadata stored using decorators. The difference lies in the full path, which should be detected for the given file:

Output of `console.log(MetadataStorage.getMetadata())` using Bun:

```javascript
{
  "Book-2b1f94ef23b79bf90eb891cae1df7a90": EntityMetadata {
    // ...
    className: "Book",
    // path should be the full path to the file
    path: "Book",
    class: [class Book],
    name: "Book",
  },
}
```

When it is working with Node, the output looks like this:

 ```javascript
 {
  'Book-1416ad576e3bd48bd0aba85e4307ae79': EntityMetadata {
    // ...
    className: 'Book',
    // full path to the file
    path: '/home/ruby/workspace/bun-playground/src/entities/Book.ts',
    class: [class Book extends BaseEntity],
    name: 'Book'
  }
}

 ```
 
This path is then used to add sources for ts-morph [here](https://github.com/mikro-orm/mikro-orm/blob/master/packages/reflection/src/TsMorphMetadataProvider.ts#L227C73-L227C79). That's the reason why it cannot find the source. Before our changes, global metadata was also added with the correct path from discovery, so everything worked despite the wrong path used by decorators.

Also this problem manifests when you do not use folder based discovery but reference the entity directly like this:

```diff
  import { TsMorphMetadataProvider } from "@mikro-orm/reflection";
  import { defineConfig, FlushMode } from "@mikro-orm/postgresql";
+ import { Book } from "./entities/Book";

  export const config = defineConfig({
    debug: true,
-   entities: ["./src/entities"],
+   entities: [Book],
    clientUrl: process.env.DATABASE_URI ??"postgresql://postgres@localhost:5432/bun-playground",
    password: "",
    metadataProvider: TsMorphMetadataProvider,
    flushMode: FlushMode.COMMIT,
    migrations: {
      snapshot: false,
    },
  });
  export default config;
```
So I think it did not work even before our changes with Bun when passing entities directly bacause paths were not correctly detected for referenced entities.

I have tried to play around and try to fix that in several places. It seems like the correct fix should address the detection of the entity path in decorators for Bun.

**Reply by co-sic** [reply]
Thanks for testing this! I used to reference entities directly with node before, but this was not working with bun so i switched to folder based discovery. I actually prefer importing the entities directly, so this was more of a workaround for bun using folder based discovery. Since it seems not to be a node problem i guess then this is my problem for using bun when its not officialy supported :/ @B4nan supporting bun in the near future is currently not a plan, right?

EDIT: Of course bun should already be supported because it says its a drop in replacement for node, so it should work without the need of any modification in the library itself. The author himself says that if there is an issue with bun that should be filed in the bun repo and not the library repo. But having field some issues myself there i think this will take a lot of time to be considered, since there are a lot of other open issues and they need to prioritize.

**Reply by Ruby184** [reply]
@co-sic @B4nan I think I can try to fix that and create a pull request. It's an issue with the 'magic' detection of the path from the error stack trace, so it's a problem with MikroORM, not Bun itself.

**Reply by co-sic** [reply]
Okay, that would be awesome of course! :tada:

**Reply by B4nan** [reply]
Will have a fix later today, my battery died when I finally cracked it :)

edit: will need a bit more time since the solution I came up with would break other cases without bun...

**Reply by B4nan** [reply]
Here you go: 6683bcc13561f009189e78325b9e9e476f452524

**Reply by Ruby184** [reply]
@B4nan I made it slightly different but your solution will fail for this test case:

```typescript
    // using bun build for generating a standalone binary will ignore the path as there is no reflect-metadata in the stack trace
    expect(Utils.lookupPathFromDecorator('Book', [
      'Error',
      '    at lookupPathFromDecorator (/$bunfs/root/app-build:40282:33)',
      '    at getMetadataFromDecorator (/$bunfs/root/app-build:52487:57)',
      '    at <anonymous> (/$bunfs/root/app-build:55177:71)',
      '    at DecorateProperty (/$bunfs/root/app-build:41850:36)',
      '    at __legacyDecorateClassTS (/$bunfs/root/app-build:50:25)',
      '    at module code (/$bunfs/root/app-build:284492:24)',
      '    at moduleEvaluation (:1:11)',
      '    at moduleEvaluation (:1:11)',
      '    at <anonymous> (:2:1)',
    ])).toBe('Book');
```
so I implemented it as follows and wanted to PR later :)

```typescript
    if (line === -1) {
      // here we handle bun which stack is different from nodejs so we search for reflect-metadata
      const reflectLine = stack.findIndex(line => Utils.normalizePath(line).includes('node_modules/reflect-metadata/Reflect.js'));

      if (reflectLine === -1 || reflectLine + 2 >= stack.length || !stack[reflectLine + 1].includes('bun:wrap')) {
        return name;
      }

      line = reflectLine + 2;
    }
```

**Reply by B4nan** [reply]
Oh, nice, would you PR that now, or should I? 

(I want to release a new stable nowish)

**Reply by Ruby184** [reply]
You can use the code above :) I will PR later just perf improvement for detection so it will always run only once :)

**Reply by co-sic** [reply]
Nice, thanks for the quick fix for this! :) @B4nan @Ruby184

**Reply by co-sic** [reply]
Short update for anybody else stumbling on this: I noticed it was not working on js files, only on ts files. In my docker setup i still had a build step and run the .js files in production. But since bun itself says there is no significant downside to directly running ts files in production (see here: https://bun.sh/docs/runtime/typescript), i changed it and everything works now.

**Reply by SnowMarble** [reply]
@co-sic I think we have to account for this. There could be teams or people who use js for legacy systems or change their runtime to Bun.
